app/com/linkedin/drelephant/exceptions/ExceptionFinder.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.log4j.Logger;
+
+
+public class ExceptionFinder {
+  private final Logger logger = Logger.getLogger(ExceptionFinder.class);
+  private HadoopException _exception;
+  private AzkabanClient _azkabanClient;
+  private MRClient _mrClient;
+  private String azkabanLogOffset = "0";
+  private String azkabanLogLengthLimit = "99999999";
+
+  public ExceptionFinder(String url)
+      throws URISyntaxException {
+
+    String execId = null;
+    List<NameValuePair> params = URLEncodedUtils.parse(new URI(url), "UTF-8");
+    for (NameValuePair param : params) {
+      if (param.getName() == "execid") {
+        execId = param.getValue();
+      }
+    }
+
+    _mrClient = new MRClient();
+    _azkabanClient = new AzkabanClient(url);
+    _azkabanClient.azkabanLogin(username, password);
+    String rawFlowLog = _azkabanClient.getAzkabanFlowLog(azkabanLogOffset, azkabanLogLengthLimit);
+
+    _exception = analyzeAzkabanFlow(execId, rawFlowLog);
+  }
+
+  private HadoopException analyzeAzkabanFlow(String execId, String rawAzkabanFlowLog) {
+    HadoopException flowLevelException = new HadoopException();
+    List<HadoopException> childExceptions = new ArrayList<HadoopException>();
+
+    AzkabanFlowLogAnalyzer analyzedLog = new AzkabanFlowLogAnalyzer(rawAzkabanFlowLog);
+    Set<String> failedAzkabanJobIds = analyzedLog.getFailedSubEvents();
+
+    for (String failedAzkabanJobId : failedAzkabanJobIds) {
+      String rawAzkabanJobLog =
+          _azkabanClient.getAzkabanJobLog(failedAzkabanJobId, azkabanLogOffset, azkabanLogLengthLimit);
+      HadoopException azkabanJobLevelException = analyzeAzkabanJob(failedAzkabanJobId, rawAzkabanJobLog);
+      childExceptions.add(azkabanJobLevelException);
+    }
+
+    flowLevelException.setType(HadoopException.HadoopExceptionType.FLOW);
+    flowLevelException.setId(execId);
+    flowLevelException.setLoggingEvent(null); // No flow level exception
+    flowLevelException.setChildExceptions(childExceptions);
+    return flowLevelException;
+  }
+
+  private HadoopException analyzeAzkabanJob(String azkabanJobId, String rawAzkabanJobLog) {
+    HadoopException azkabanJobLevelException = new HadoopException();
+    List<HadoopException> childExceptions = new ArrayList<HadoopException>();
+    AzkabanJobLogAnalyzer analyzedLog = new AzkabanJobLogAnalyzer(rawAzkabanJobLog);
+    Set<String> mrJobIds = analyzedLog.getSubEvents(); // returns all mrjobs in the azkaban job
+
+    for (String mrJobId : mrJobIds) {
+      //To do: Check if mr job logs are there or not in job history server
+      String rawMRJobLog = _mrClient.getMRJobLog(mrJobId);
+      if (rawMRJobLog != null) {
+        //To do: rawMRJob is null for successful mr jobs but this is not a job to figure out whether a job failed or succeeded
+        HadoopException mrJobLevelException = analyzeMRJob(mrJobId, rawAzkabanJobLog);
+        childExceptions.add(mrJobLevelException);
+      }
+    }
+    if (!childExceptions.isEmpty()) {
```

Comment:
Add some comments on why you want to look at the job level exception only if childExceptions is empty